vignettes/unit_test_guidance.Rmd

```diff
@@ -0,0 +1,151 @@
+them and introduce bugs. Test cases could be used as a form of documentation, so
+they obviously need to be readable.
+
+### Running Unit Tests Part of the Build Process
+Automate the whole process of running the unit tests and taking some action when
+they fail. Your build process should execute your unit tests and mark the build 
+as broken when the tests fail.
+
+# Writing Unit Tests in Admiral
+
+## Plan your Unit Tests
+Start by considering the derivation rule you are testing and the possible 
+arguments/flexibilities of your function code. Then plan which scenarios you will
+test. These can either involve generating different input test cases or feeding 
+them into different calls of your function.
+
+## Tests Should be Robust to Cover Realistic Data Scenarios
+For generating input test cases, it can be helpful to consider regular cases 
+(expected common data scenarios), boundary cases (where data points are close or
+equal), and special cases (uncommon but valid data scenarios, e.g. missing or 
+special characters). Although you will never cover every single eventuality of 
+possible input data (no reliability testing method ever gives 100% certainty), 
+you do need to give confidence that the code is robust enough to work across most
+data scenarios.
+
+## Testing Should Cover Possible Arguments
+For the different calls of your function, consider how the user might apply your
+function and test a variety of possible calls, whilst still remembering the tips
+above that tests should be fast and simple. 
+This is only needed in cases where the complexity and level of flexibility of 
+your function justifies it, e.g. see the test script: https://github.com/Roche-GSK/admiral/blob/master/tests/testthat/test-derive_extreme_flag.R.
+
+## Exported Functions
+Don’t forget to add a unit test for each exported function.
+
+## Set up the Test Script
+Within {admiral} folder https://github.com/Roche-GSK/admiral/tree/master/tests/testthat, 
+add a script with the naming convention “test-<function name>.R” and use the 
+following format:
+
+```
+
+context("test-<function name>")
+
+test_that("<function name> Test 1: <Explanation of the test>", {
+  input <- tibble::tribble(
+    ~inputvar1, ~inputvar2, ...
+    <Add Test Data Scenarios>
+    ...
+  )
+
+  expected_output <- mutate(input, outputvar = c(<Add Expected Outputs>))
+
+  expect_equal(<function name>(input), expected_output)
+})
+```
+Ensure you give a meaningful explanation of the test in the testthat call, as 
+these will be compiled in the package validation report. Having the name of the
+function and test ID included in title will also help with tracability.
+
+Once you have tested your unit test program, you can run all unit tests from
+the console, as follows.
+
+```
+devtools::check
+```
+
+## Automation of Unit Tests
+
+When a user actions a pull request in admiral github repo, the unit tests are 
+automatically run and pull request will be denied if any unit tests fail.
```

Comment:
Please write as **G**it**H**ub everywhere.

Reply:
Done